File: runtime/src/main/java/com/siimkinks/sqlitemagic/SelectionTable.java
package com.siimkinks.sqlitemagic;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.siimkinks.sqlitemagic.internal.SimpleArrayMap;

import java.util.ArrayList;
import java.util.Collections;

final class SelectionTable<T> extends Table<T> {
  @NonNull
  private final String compiledSql;
  @Nullable
  private final String[] args;
  @NonNull
  private final String[] observedTables;
  private final Query.Mapper<T> mapper;

  private SelectionTable(@NonNull String compiledSql,
                         @Nullable String alias,
                         @Nullable String[] args,
                         @NonNull String[] observedTables,
                         Query.Mapper<T> mapper) {
    super("", alias, 0);
    this.compiledSql = compiledSql;
    this.args = args;
    this.mapper = mapper;
    this.observedTables = observedTables;
  }

  @SuppressWarnings("unchecked")
  static <T> SelectionTable<T> from(@NonNull SelectBuilder<?> selectTableBuilder,
                                    @Nullable String alias) {
    final String compiledSql;
    final String[] args;
    final Query.Mapper<T> mapper;
    final String[] observedTables;
    final CompiledSelect<Object, ?> compiledSelect = selectTableBuilder.build();
    if (compiledSelect instanceof CompiledSelect1Impl<?, ?>) {
      final CompiledSelect1Impl cs = (CompiledSelect1Impl) compiledSelect;
      compiledSql = cs.sql;
      args = cs.args;
      mapper = cs.mapper;
      observedTables = cs.observedTables;
    } else if (compiledSelect instanceof CompiledSelectImpl<?, ?>) {
      final CompiledSelectImpl cs = (CompiledSelectImpl) compiledSelect;
      compiledSql = cs.sql;
      args = cs.args;
      mapper = cs.mapper;
      observedTables = cs.observedTables;
    } else throw new RuntimeException("Unknown compiled select");
    return new SelectionTable<>(compiledSql, alias, args, observedTables, mapper);
  }

  void linkWithOuterSelect(@NonNull SelectBuilder<?> outerSelectBuilder) {
    if (args != null) {
      Collections.addAll(outerSelectBuilder.args, args);
    }
  }

  @Override
  void appendToSqlFromClause(@NonNull StringBuilder sb) {
    sb.append('(');
    sb.append(compiledSql);
    sb.append(')');
    if (hasAlias) {
      sb.append(" AS ")
          .append(alias);
    }
  }

  @Override
  boolean perfectSelection(@NonNull ArrayList<String> observedTables, @Nullable SimpleArrayMap<String, String> tableGraphNodeNames, @Nullable SimpleArrayMap<String, Integer> columnPositions) {
    final String[] subqueryObservedTables = this.observedTables;
    final int length = subqueryObservedTables.length;
    for (int i = 0; i < length; i++) {
      final String tableName = subqueryObservedTables[i];
      if (!observedTables.contains(tableName)) {
        observedTables.add(tableName);
      }
    }
    return false;
  }

  @NonNull
  @Override
  Query.Mapper<T> mapper(@Nullable SimpleArrayMap<String, Integer> columnPositions, SimpleArrayMap<String, String> tableGraphNodeNames, boolean queryDeep) {
    return mapper;
  }

  @NonNull
  @Override
  public Table<T> as(@NonNull String alias) {
    return new SelectionTable<>(compiledSql, alias, args, observedTables, mapper);
  }
}

File: runtime/src/main/kotlin/com/siimkinks/sqlitemagic/SelectionTable.kt
package com.siimkinks.sqlitemagic

import com.siimkinks.sqlitemagic.internal.SimpleArrayMap

internal class SelectionTable<T> private constructor(
  private val compiledSql: String,
  alias: String?,
  private val args: Array<String?>?,
  private val observedTables: Array<String>,
  private val rowMapper: Query.Mapper<T>
) : Table<T>(
  name = "",
  alias = alias,
  nrOfColumns = 0,
  mapper = { _, _, _ -> rowMapper }
) {
  internal fun linkWithOuterSelect(outerSelectBuilder: SelectBuilder<*>) {
    args?.toList()?.let(outerSelectBuilder.args::addAll)
  }

  override fun appendToSqlFromClause(sb: StringBuilder) {
    sb.append('(')
      .append(compiledSql)
      .append(')')
    if (hasAlias) {
      sb.append(" AS ")
        .append(alias)
    }
  }

  override fun perfectSelection(
    observedTables: ArrayList<String>,
    tableGraphNodeNames: SimpleArrayMap<String, String>?,
    columnPositions: SimpleArrayMap<String, Int>?
  ): Boolean {
    this.observedTables.forEach { tableName ->
      if (tableName !in observedTables) {
        observedTables.add(tableName)
      }
    }
    return false
  }

  override fun `as`(alias: String): Table<T> = SelectionTable(
    compiledSql = compiledSql,
    alias = alias,
    args = args,
    observedTables = observedTables,
    rowMapper = rowMapper
  )

  companion object {
    @Suppress("UNCHECKED_CAST")
    internal fun <T> from(
      selectTableBuilder: SelectBuilder<*>,
      alias: String?
    ): SelectionTable<T> {
      val compiledSelect = selectTableBuilder.build<T>()
      val details = (compiledSelect as? CompiledSelectDetails)
        ?: error("Unknown compiled select")
      val mapper = (compiledSelect as? Query.DatabaseQuery<List<T>, T>)?.mapper
      return SelectionTable(
        compiledSql = details.sql,
        alias = alias,
        args = details.args,
        observedTables = details.observedTables,
        rowMapper = checkNotNull(mapper)
      )
    }
  }
}
